### ISSUE TITLE
Could not import extension sphinx.builders.linkcheck (exception: urllib3 v2.0 only supports OpenSSL 1.1.1+)

### ISSUE THREAD

**Issue by stereoshoots** [readthedocs/readthedocs.org#10290]
## Details

* Read the Docs project URL: https://readthedocs.org/projects/gitlab-board
* Build URL (if applicable): https://readthedocs.org/projects/gitlab-board/builds/20387388/
* Read the Docs username (if applicable): https://readthedocs.org/profiles/stereoshoots/

## Expected Result

New revision of docs should be builded.

## Actual Result

Error is thrown:

```
Could not import extension sphinx.builders.linkcheck (exception: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with OpenSSL 1.0.2n  7 Dec 2017. See: https://github.com/urllib3/urllib3/issues/2168)
```

Is this a problem on my side? It installs pip packages in build, how should i fix this?

**Comment by agjohnson** [comment]
More specifically, to use a configuration file, you should use a configuration file that that specifies the build OS:

```yaml
version: 2

# Set the version of Python and other tools you might need
build:
  os: ubuntu-22.04
```

This will ensure you are not using one of the older images.

**Comment by agjohnson** [comment]
@sethmlarson Thanks for the note, I'd agree the configuration fix should be the recommended fix. I'll edit the guidance above.

It seems like we're probably going to aim to get some sort of fix out here, as this is causing a bit of trouble for folks. It seems the trouble is an old pyenv compiled Python in an image we're trying to deprecate, but should be something we can fix.

Unfortunately, we're going into Write the Docs conference time, so we haven't settled on a timeframe yet.

**Comment by agjohnson** [comment]
Also, I opened an issue yesterday for updating and fixing the legacy build image:
https://github.com/readthedocs/readthedocs-docker-images/issues/194

We'll track updating and fixing the image there. I'll see what I can do today and tomorrow

**Comment by mmmarinho** [comment]
If another poor soul is here from using the [readthedocs template](https://github.com/readthedocs/tutorial-template), here's what worked for me.

Add the `.readthedocs.yaml` at the root of your repo. Note the `docs/source/conf.py`, which can be easily overlooked from the [config file example](https://docs.readthedocs.io/en/stable/config-file/v2.html).

``` yaml
# .readthedocs.yaml
# Read the Docs configuration file
# See https://docs.readthedocs.io/en/stable/config-file/v2.html for details

# Required
version: 2

# Set the version of Python and other tools you might need
build:
  os: ubuntu-22.04
  tools:
    python: "3.11"
    # You can also specify other tool versions:
    # nodejs: "19"
    # rust: "1.64"
    # golang: "1.19"

# Build documentation in the docs/ directory with Sphinx
sphinx:
   configuration: docs/source/conf.py

# If using Sphinx, optionally build your docs in additional formats such as PDF
# formats:
#    - pdf

# Optionally declare the Python requirements required to build your docs
# python:
#  install:
#   - requirements: docs/requirements.txt
```

**Comment by andremann** [comment]
It's not very clear how to hotfix this.
I am relatively new to readthedocs and I tried to create a `.readthedocs.yaml` file as indicated in some comments above, but I could manage to make it build again. The error changes, but I can't get it to build.

**Comment by andremann** [comment]
Thanks @mmmarinho! You saved my day! I didn't realise it was `docs/conf.py` and not`docs/source/conf.py`

**Comment by stereoshoots** [comment]
> Hi, this is because urlib3 was released yesterday, and that version isn't compatible with the docker image you are using. To use a more recent version of our docker images, please use a config file (https://docs.readthedocs.io/en/stable/config-file/v2.html), ~or pin urllib3 in your requirements to a previous version.~
> 
> Edit: projects should use a configuration file and leave urllib3 unpinned

It worked after adding `.readtthedocs.yaml` and using recommended image and OS. Anyways, i have a slight problem after doing it. My navbar is now empty, without headers.

<img width="301" alt="image" src="https://user-images.githubusercontent.com/3724558/236401712-106e1845-112a-46ec-9e98-f98ed626807d.png">

I believe, i'm missing now some other configurations.

Thanks!

**Comment by natt96z** [comment]
> If another poor soul is here from using the [readthedocs template](https://github.com/readthedocs/tutorial-template), here's what worked for me.
> 
> Add the `.readthedocs.yaml` at the root of your repo. Note the `docs/source/conf.py`, which can be easily overlooked from the [config file example](https://docs.readthedocs.io/en/stable/config-file/v2.html).
> 
> ```yaml
> # .readthedocs.yaml
> # Read the Docs configuration file
> # See https://docs.readthedocs.io/en/stable/config-file/v2.html for details
> 
> # Required
> version: 2
> 
> # Set the version of Python and other tools you might need
> build:
>   os: ubuntu-22.04
>   tools:
>     python: "3.11"
>     # You can also specify other tool versions:
>     # nodejs: "19"
>     # rust: "1.64"
>     # golang: "1.19"
> 
> # Build documentation in the docs/ directory with Sphinx
> sphinx:
>    configuration: docs/source/conf.py
> 
> # If using Sphinx, optionally build your docs in additional formats such as PDF
> # formats:
> #    - pdf
> 
> # Optionally declare the Python requirements required to build your docs
> # python:
> #  install:
> #   - requirements: docs/requirements.txt
> ```

Really saved my bacon, thanks!

**Comment by agjohnson** [comment]
@stereoshoots check your build output for errors/warning in the Sphinx build command. The upgrade to the operating system image also means there were a number of dependencies changed in the process.

@RemDelaporteMathurin it's not just `version: 2` that is required to avoid the bug here, but specifically `build.os: ubuntu-22.04`.

**Comment by BeyondEvil** [comment]
> @stsewd I suspect this issue will get a lot of traffic, if possible could the correct answer w/ example of upgrading to `build.os: ubuntu-22.04` instead of pinning urllib3 be elevated to the first reply. I suspect folks may be looking for quick solutions to this incompatibility and I'd like people to keep urllib3 unpinned if it's not blocked by anything else in their environment.


Does this mean that the instructions [here](https://docs.readthedocs.io/en/latest/guides/reproducible-builds.html#pin-your-transitive-dependencies) are no longer correct? 

Because following those instructions will pin urllib3.

**Comment by humitos** [comment]
Hi! Thank you all for reporting this issue and for your patience 😄 . 

TLDR; new Docker images with SSL 1.1.1 were deployed and they are ready to use. If your project was failing because of this, it should not be failing anymore. If you still have this problem or you find another one that looks related to the rebuild of the images, please open a new issue.

Here I'm writing down the news about this problem:

### The root cause

Read the Docs still supports some old Docker images [^1] that contains Python versions compiled with SSL 1.0.x and this SSL version was dropped support from `urlib3` in its latest release. `urllib3` is used by `requests` behind the scenes and `requests` is used by _a lot of things_; in particular Sphinx `linkcheck` command where this was originally found.

### The workaround

Upgrade your `.yaml` file to use `build.os: ubuntu-22.04` as mentioned in https://github.com/readthedocs/readthedocs.org/issues/10290#issuecomment-1535120995

Besides being a workaround, this upgrade is useful to be up-to-date with the latest version of the RTD config file, which introduces you to some nice features like [build customization](https://docs.readthedocs.io/en/latest/build-customization.html)

### The solution

We re-built the Docker images so they use SSL 1.1.1 to compile the Python versions via `pyenv`. This should be transparent for all the projects and they should not notice any new error, 🤞🏼 

### The future

SSL 1.1.1 is gonna be unsupported later this year as well, based on https://github.com/urllib3/urllib3/issues/2168; so, we are thinking about deprecating and removing these Docker images (and the whole `build.image` key) completely before that time. We need to keep talking about this and how to do it without breaking lot of projects.


[^1]: this is `build.image` key in the config file with any of the values `stable`, `latest` and `testing`. A lot of users were hitting this because `build.image: latest` with Python 3.7 is still the default; but we are working on changing this default to a more up-to-date version of Ubuntu and Python as well. Please, if you are using `build.image` in your `.yaml` file, upgrade it to `build.os: ubuntu-22.04` as mentioned in  https://github.com/readthedocs/readthedocs.org/issues/10290#issuecomment-1535120995

**Comment by agjohnson** [comment]
@BeyondEvil it's still recommended general to use an automated process to explicitly pin all of your direct and transitive dependencies.

But on the other hand, I wouldn't recommend that solution as a fix here either. This requires additional tooling like pip-tools to automate. For the basic user, the direct fix above is the easiest first step, but defining dependencies explicitly is a great idea still

**Comment by BeyondEvil** [comment]
> @BeyondEvil it's still recommended general to use an automated process to explicitly pin all of your direct and transitive dependencies.
> 
> But on the other hand, I wouldn't recommend that solution as a fix here either. This requires additional tooling like pip-tools to automate. For the basic user, the direct fix above is the easiest first step, but defining dependencies explicitly is a great idea still

Ok, my project has a super old RTD (and Sphinx) setup - so this prompted me to update at least parts of it.

So I've added the `.readthedocs.yaml` config file and the `requirements.in` and (generated by pip-compile) `requirements.txt` files.

I should be fine then?

**Comment by agjohnson** [comment]
@BeyondEvil Correct! If you do add a pinned dependency your project should indeed be okay, especially if you have followed the guide for reproducible builds.

**Comment by humitos** [comment]
We haven't heard people hitting any new issue after the deploy with the fix we did yesterday; so I'm going to close this issue.

If you are arrive at this issue by browsing the internet, please refer to this comment in particular https://github.com/readthedocs/readthedocs.org/issues/10290#issuecomment-1542841524

Thanks you all for the support! ❤️

**Comment by fjarri** [comment]
Does it mean that the workaround with the explicit `build.os: ubuntu-22.04` is no longer necessary? It will Just Work (TM)?

**Comment by humitos** [comment]
@fjarri using `build.os: ubuntu-22.04` plus `build.tools.python` is the preferred and supported way to use Read the Docs. This is the way to go 👍🏼 

Note that not having a `readthedocs.yaml` in your project or using `build.image: latest` is deprecated and it will removed in the near future. We will be communicating this more clearly to projects that are still using it in the next weeks.

**Comment by ExtremeAddons** [comment]
> We haven't heard people hitting any new issue after the deploy with the fix we did yesterday; so I'm going to close this issue.
> 
> If you are arrive at this issue by browsing the internet, please refer to this comment in particular [#10290 (comment)](https://github.com/readthedocs/readthedocs.org/issues/10290#issuecomment-1542841524)
> 
> Thanks you all for the support! ❤️

I was having this error for at least 2 days. Now I was able to compile the documentation without any problems. Looks like you fixed it great
